src/gpu/intel/jit/ngen/ngen_level_zero.hpp

```diff
@@ -21,13 +21,100 @@
 
 #include "gpu/intel/sycl/l0/level_zero/ze_api.h"
 
+#if defined(__linux__)
+#include <dlfcn.h>
+#elif defined(_WIN32)
+#include "windows.h"
+#else
+#error "Level Zero is supported on Linux and Windows only"
+#endif
+
 #include <sstream>
 
 #include "ngen_elf.hpp"
 #include "ngen_interface.hpp"
 
 namespace NGEN_NAMESPACE {
 
+// Dynamically loaded level_zero functions
+namespace {
+
+void *find_ze_symbol(const char *symbol) {
+#if defined(__linux__)
+    void *handle = dlopen("libze_loader.so.1", RTLD_NOW | RTLD_LOCAL);
+#elif defined(_WIN32)
+    // Use LOAD_LIBRARY_SEARCH_SYSTEM32 flag to avoid DLL hijacking issue.
+    HMODULE handle = LoadLibraryExA(
+            "ze_loader.dll", nullptr, LOAD_LIBRARY_SEARCH_SYSTEM32);
+#endif
+    if (!handle) {
+        assert(!"not expected");
+        return nullptr;
+    }
+
+    using zeInit_decl_t = ze_result_t (*)(ze_init_flags_t flags);
+    const ze_init_flags_t default_ze_flags = 0;
+#if defined(__linux__)
+    static const ze_result_t ze_result = reinterpret_cast<zeInit_decl_t>(
+            dlsym(handle, "zeInit"))(default_ze_flags);
+    void *f = reinterpret_cast<void *>(dlsym(handle, symbol));
+#elif defined(_WIN32)
+    static const ze_result_t ze_result = reinterpret_cast<zeInit_decl_t>(
+            GetProcAddress(handle, "zeInit"))(default_ze_flags);
+    void *f = reinterpret_cast<void *>(GetProcAddress(handle, symbol));
+#endif
+
+    if (ze_result != ZE_RESULT_SUCCESS) assert(!"not expected");
+    if (!f) assert(!"not expected");
+    return f;
+}
+
+template <typename F>
+F find_ze_symbol(const char *symbol) {
+    return (F)find_ze_symbol(symbol);
+}
+
+ze_result_t func_zeModuleCreate(ze_context_handle_t hContext,
+        ze_device_handle_t hDevice, const ze_module_desc_t *desc,
+        ze_module_handle_t *phModule,
+        ze_module_build_log_handle_t *phBuildLog) {
+    static auto f = find_ze_symbol<decltype(&zeModuleCreate)>("zeModuleCreate");
+
+    if (!f) return ZE_RESULT_ERROR_UNINITIALIZED;
+    return f(hContext, hDevice, desc, phModule, phBuildLog);
+}
```

Comment:
Could you reduce the boilerplate with some generic code + macros?

```
#define ZE_INDIRECT_API(f) \
    template <typename... Args> ze_result_t call_##f(Args&&... args) { \
        static auto f = find_ze_symbol<decltype(%f)>(#f);
        if (!f) return ZE_RESULT_ERROR_UNINITIALIZED;
        return f(std::forward<Args>(args)...);
    }

ZE_INDIRECT_API(zeModuleCreate)
ZE_INDIRECT_API(zeModuleDestroy)
/* ... */
```

Reply:
Done, thanks!